src/NUnitEngine/nunit.engine.core/Services/ExtensionManager.cs

```diff
@@ -118,13 +120,19 @@ public virtual void FindExtensionPoints(params Assembly[] targetAssemblies)
         /// <inheritdoc/>
         public void FindExtensions(string startDir)
         {
-            // Create the list of possible extension assemblies,
-            // eliminating duplicates, start in the provided directory.
-            FindExtensionAssemblies(_fileSystem.GetDirectory(startDir));
+            // Ignore a call for a directory we have already used
+            if (!_extensionDirectories.Contains(startDir))
+            {
+                _extensionDirectories.Add(startDir);
 
-            // Check each assembly to see if it contains extensions
-            foreach (var candidate in _assemblies)
-                FindExtensionsInAssembly(candidate);
+                // Create the list of possible extension assemblies,
+                // eliminating duplicates, start in the provided directory.
+                FindExtensionAssemblies(_fileSystem.GetDirectory(startDir));
+
+                // Check each assembly to see if it contains extensions
+                foreach (var candidate in _assemblies)
+                    FindExtensionsInAssembly(candidate);
```

Comment:
`_assemblies` is appended to in each call to `FindExtensionAssemblies`.
This means that one the 2nd call to `FindExtensions` for a 2nd directory, we will revisit the assemblies found in the 1st call again. This could mean that the same extension is loaded multiple times.
In addition the assemblies loaded by `FindStandardExtensions` will be visited again.

If `FindExtensionsAssemblies` would return only the list of new assemblies, only those would be searched.

Alternative split the finding of assemblies from the finding of extension in those assemblies.
Delete this method and make `FindExtensionAssemblies` public.
Then after all assemblies are found call a new method `FindExtensionsInAssemblies()` which iterates over `_assemblies` like here.
In addition remove the iteration from `FindStandardExtensions`.

Reply:
We keep a dictionary of all assembly paths which have already been seen so that we don't re-process them. This was needed before we changed to this three-stage process of finding extensions, because the user could reference the same directory or assembly in multiple `.addins` files. You can see this in the `ProcessDirectory` and `ProcessCandidateAssembly` methods.

Please look at that code to confirm that it works as intended. Meanwhile, I'll try to construct a test to prove it and prevent regression.

By the way, the new check on _extensionDirectories wasn't strictly necessary, but I wanted to have a clean, non-duplicated list of extension directories for use in this issue. In the end, I didn't make use of it as intended, but I have kept it for a future enhancement.

Reply:
Indeed you don't visit the same directory twice (_existingDirectories), 
don't process the same directory multiple times (_visited) and 
don't add the same assembly multiple times (IsDuplicateOf),
However, unless I missed some other guard, I still think you load extensions multiple times.

Say we have two directories on the search lists: A and B, each of them has 1 assembly with 1 extension.
When calling FindExtensions(A) it adds A/Assembly to `_assemblies` and calls 'FindExtensionsInAssembly(A/Assemby)'
When calling FindExtensions(B) it adds B/Assembly to `_assemblies` and calls both FindExtensionsInAssembly(A/Assemby) and FindExtensionsInAssembly(A/Assemby)
There is no protection in FindExtensionsInAssembly.

If you have a setup with 2 extension, but them in separate directories and check the logs.
There is no test on `FindExtensions`, it would require creating extensions on disk and not the Fake ones.

Reply:
What we need to demonstrate is as follows.
1. There is only one place where an assembly can be added to _assemblies: `ProcessCandidateAssembly`.
2. That method could be called for a duplicate assembly if...
    2.1 The same asssembly path is specified in two .addins files
    2.2 The same assembly is found in different directories.
3. In either case, it should not add the duplicate to the list...
    3.1 The same path should be ruled out by `WasVistited`.
    3.2 The same AssemblyName should be ruled out by `IsDuplicateOf`.

At this point, I'm not convinced it's happening as it should, so I'll set up extensions in two directories and check the logs as you suggest. But really, all of the above needs tests. I think we can create unit tests without anything on disk because we aren't dealing with a list of assemblies here, but a list of `ExtensionAssembly` and that class can be faked.

Interesting problem. :-)

Reply:
I'm fairly convinced we are avoiding actual duplicates but we're definitely doing too much re-processing of the assembly list. I'm going to use your first suggestion of returning only the list of new assemblies. The second approach means exposing FindExtensionAssemblies, which should really be called FindPossibleExtensionAssemblies or something like that. I don't think that's a good public method.